File: AutoUpdateJS/src/services/VersionProvider.js
import { VersionInfo } from '../models/VersionInfo.js';
import { httpClient } from '../utils/httpClient.js';
import { readJsonFile, fileExists } from '../utils/fileUtils.js';
import logger from '../utils/logger.js';

/**
 * Service for providing version information from local files and GitHub.
 */
export class VersionProvider {
  /**
   * Creates a new VersionProvider instance.
   * @param {AutoUpdateConfig} config - Configuration instance
   */
  constructor(config) {
    this.config = config;
    this.httpClient = new httpClient.constructor({
      timeout: config.httpTimeout
    });
  }

  /**
   * Gets the local version information.
   * @returns {Promise<VersionInfo>} Local version info
   */
  async getLocalVersion() {
    try {
      if (!(await fileExists(this.config.versionFilePath))) {
        logger.warn(`Version file not found at ${this.config.versionFilePath}, returning default version`);
        return new VersionInfo({ version: '0.0.0' });
      }

      const versionData = await readJsonFile(this.config.versionFilePath);
      const versionInfo = new VersionInfo(versionData);

      logger.info(`Local version loaded: ${versionInfo.version}`);
      return versionInfo;
    } catch (error) {
      logger.error(`Error reading local version file: ${error.message}`, error);
      return new VersionInfo({ version: '0.0.0' });
    }
  }

  /**
   * Gets the remote version information from GitHub.
   * @param {string} owner - Repository owner
   * @param {string} repo - Repository name
   * @param {string} branch - Branch name
   * @returns {Promise<VersionInfo|null>} Remote version info or null if error
   */
  async getRemoteVersion(owner, repo, branch) {
    try {
      const versionUrl = `${this.config.getRawContentUrl(branch)}/${this.config.versionFilePath}`;
      logger.info(`Fetching remote version from: ${versionUrl}`);

      const response = await this.httpClient.get(versionUrl);

      if (response.statusCode !== 200) {
        logger.warn(`Failed to fetch remote version: HTTP ${response.statusCode}`);
        return null;
      }

      const versionInfo = new VersionInfo(response.body);
      logger.info(`Remote version fetched: ${versionInfo.version}`);

      return versionInfo;
    } catch (error) {
      logger.error(`Error fetching remote version: ${error.message}`, error);
      return null;
    }
  }

  /**
   * Gets the remote version for the configured production branch.
   * @returns {Promise<VersionInfo|null>} Remote version info
   */
  async getProductionVersion() {
    return this.getRemoteVersion(
      this.config.githubOwner,
      this.config.githubRepo,
      this.config.productionBranch
    );
  }

  /**
   * Gets the remote version for the configured development branch.
   * @returns {Promise<VersionInfo|null>} Remote version info
   */
  async getDevelopmentVersion() {
    return this.getRemoteVersion(
      this.config.githubOwner,
      this.config.githubRepo,
      this.config.developmentBranch
    );
  }

  /**
   * Compares local and remote versions.
   * @param {VersionInfo} localVersion - Local version
   * @param {VersionInfo} remoteVersion - Remote version
   * @returns {number} 1 if remote is newer, -1 if local is newer, 0 if equal
   */
  compareVersions(localVersion, remoteVersion) {
    return remoteVersion.compareTo(localVersion);
  }

  /**
   * Checks if an update is available.
   * @returns {Promise<{hasUpdate: boolean, localVersion: VersionInfo, remoteVersion: VersionInfo|null}>}
   */
  async checkForUpdate() {
    const localVersion = await this.getLocalVersion();
    const remoteVersion = await this.getProductionVersion();

    if (!remoteVersion) {
      return {
        hasUpdate: false,
        localVersion,
        remoteVersion: null
      };
    }

    const hasUpdate = remoteVersion.isGreaterThan(localVersion);

    return {
      hasUpdate,
      localVersion,
      remoteVersion
    };
  }
}

File: AutoUpdatePython/autoupdate/version_provider.py
"""
Version provider for GitHub-based version management.
"""

import json
import urllib.request
import urllib.error
from typing import Optional
from .core import VersionInfo
from .config import AutoUpdateConfig


class VersionProvider:
    """
    Service for providing version information from local files and GitHub.
    """

    def __init__(self, config: AutoUpdateConfig):
        """
        Initialize VersionProvider.

        Args:
            config: AutoUpdateConfig instance
        """
        self.config = config

    def get_local_version(self) -> VersionInfo:
        """
        Get the local version information from version.json file.

        Returns:
            VersionInfo instance
        """
        try:
            with open(self.config.version_file_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
                return VersionInfo.from_dict(data)
        except (FileNotFoundError, json.JSONDecodeError, KeyError):
            # Return default version if file doesn't exist or is invalid
            return VersionInfo(version="0.0.0")

    def get_remote_version(self, owner: str, repo: str, branch: str) -> Optional[VersionInfo]:
        """
        Get remote version information from GitHub.

        Args:
            owner: GitHub repository owner
            repo: GitHub repository name
            branch: Branch name

        Returns:
            VersionInfo instance or None if error
        """
        try:
            version_url = f"{self.config.get_raw_content_url(branch)}/{self.config.version_file_path}"

            with urllib.request.urlopen(version_url, timeout=self.config.http_timeout) as response:
                if response.status != 200:
                    return None

                data = json.loads(response.read().decode('utf-8'))
                return VersionInfo.from_dict(data)

        except (urllib.error.URLError, json.JSONDecodeError, KeyError):
            return None

    def get_production_version(self) -> Optional[VersionInfo]:
        """
        Get the production version from GitHub.

        Returns:
            VersionInfo instance or None if error
        """
        return self.get_remote_version(
            self.config.github_owner,
            self.config.github_repo,
            self.config.production_branch
        )

    def get_development_version(self) -> Optional[VersionInfo]:
        """
        Get the development version from GitHub.

        Returns:
            VersionInfo instance or None if error
        """
        return self.get_remote_version(
            self.config.github_owner,
            self.config.github_repo,
            self.config.development_branch
        )

    def check_for_update(self) -> dict:
        """
        Check if an update is available.

        Returns:
            Dictionary with update information:
            {
                'has_update': bool,
                'current_version': VersionInfo,
                'remote_version': VersionInfo or None
            }
        """
        current_version = self.get_local_version()
        remote_version = self.get_production_version()

        if remote_version is None:
            return {
                'has_update': False,
                'current_version': current_version,
                'remote_version': None
            }

        has_update = remote_version.is_greater_than(current_version)

        return {
            'has_update': has_update,
            'current_version': current_version,
            'remote_version': remote_version
        }
